sklearn/preprocessing/_encoders.py

```diff
@@ -673,37 +685,69 @@ def _legacy_transform(self, X):
 
         return out if self.sparse else out.toarray()
 
+    def _make_onehot_sparse_matrix(self, labels, mask, cat_ns):
+        flat_mask = mask.ravel()
+        n_values = np.array([0] + cat_ns)
+        feature_indices = np.cumsum(n_values)
+
+        indices = (labels + feature_indices[:-1]).ravel()[flat_mask]
+        indptr = mask.sum(axis=1).cumsum()
+        indptr = np.insert(indptr, 0, 0)
+        data = np.ones(len(indices))
+
+        out = sparse.csr_matrix((data, indices, indptr),
+                                shape=(len(labels), feature_indices[-1]),
+                                dtype=self.dtype)
+        return out
+
+    def _make_nan_sparse_matrix(self, Xi_missing, n_categories):
+        n_missing = np.sum(Xi_missing)
+        indptr = np.cumsum(Xi_missing) * n_categories
+        indptr = np.insert(indptr, 0, 0)
+        indices = np.full((n_missing, n_categories), np.arange(n_categories))
+        indices = indices.ravel()
+        data = np.full(len(indices), np.nan)
+        na_matrix = sparse.csr_matrix((data, indices, indptr),
+                                      (len(Xi_missing), n_categories))
+        return na_matrix
+
     def _transform_new(self, X):
         """New implementation assuming categorical input"""
         # validation of X happens in _check_X called by _transform
-        X_int, X_mask = self._transform(X, handle_unknown=self.handle_unknown)
-
-        n_samples, n_features = X_int.shape
+        X_int, X_missing, X_unknown = (
+            self._transform(X, missing_values=self.missing_values,
+                            handle_unknown=self.handle_unknown))
 
         if self.drop is not None:
             to_drop = self.drop_idx_.reshape(1, -1)
 
             # We remove all the dropped categories from mask, and decrement all
             # categories that occur after them to avoid an empty column.
 
-            keep_cells = X_int != to_drop
-            X_mask &= keep_cells
+            drop_cells = X_int == to_drop
+            X_unknown |= drop_cells
             X_int[X_int > to_drop] -= 1
-            n_values = [len(cats) - 1 for cats in self.categories_]
+            cats_ns = [len(cats) - 1 for cats in self.categories_]
```

Comment:
I don't like the name `cats_ns`. Please make it more explicit. `n_values` was okay